fuzz.go

```diff
@@ -35,3 +38,98 @@ func Fuzz(data []byte) int {
 	}
 	return 0
 }
+
+// FuzzReadSubTreesWithProof can be used by go-fuzz to test creating a merkle
+// tree from cached subTrees and creating/proving a merkle proof on this tree.
+func FuzzReadSubTreesWithProof(data []byte) int {
+	// Use the first two bytes to determine the proof index.
+	if len(data) < 2 {
+		return -1
+	}
+	// Use the first two bytes to determine the proof index.
+	index := 256*uint64(data[0]) + uint64(data[1])
+	tree := New(sha256.New())
+	tree.SetIndex(index)
+	data = data[2:]
+
+	// 32 is the length of a sha256 hash and 4 bytes are used for the height of
+	// the subTree.
+	subTreeSize := 4 + 36
+	err := tree.readSubTrees(bytes.NewReader(data), subTreeSize)
+	if err != io.ErrUnexpectedEOF {
+		return 0
+	} else if err == io.ErrUnexpectedEOF && len(data) < subTreeSize {
+		return -1
+	}
+
+	// Create and verify the proof.
+	merkleRoot, proofSet, proofIndex, numLeaves := tree.Prove()
+	if len(proofSet) == 0 {
+		// proofIndex wasn't reached while creating proof.
+		return 0
+	}
+	if !VerifyProof(sha256.New(), merkleRoot, proofSet, proofIndex, numLeaves) {
+		panic("verification failed!")
+	}
+	// Output is more interesting when there is enough data to contain the
+	// index.
+	if uint64(len(data)) > uint64(subTreeSize)*index {
+		return 1
+	}
+	return 0
+}
+
+// FuzzReadSubTreesNoProof can be used by go-fuzz to test creating a merkle
+// tree from cached subTrees.
+func FuzzReadSubTreesNoProof(data []byte) int {
+	tree := New(sha256.New())
+
+	// 32 is the length of a sha256 hash and 4 bytes are used for the height of
+	// the subTree.
+	subTreeSize := 4 + 36
+	err := tree.readSubTrees(bytes.NewReader(data), subTreeSize)
+	if err != io.ErrUnexpectedEOF {
+		return 0
+	} else if err == io.ErrUnexpectedEOF && len(data) < subTreeSize {
+		return -1
+	}
+	if tree.head != nil && tree.Root() == nil {
+		panic("root shouldn't be nil for a non-empty tree")
+	}
+	// The data is better if it contains many subTrees.
+	if len(data) > 100*subTreeSize {
+		return 1
+	}
+	return 0
+}
+
+// readSubTrees is a helper function that maps the data from a io.Reader to
+// subTrees and adds them to a tree.
+func (t *Tree) readSubTrees(r io.Reader, subTreeSize int) error {
+	if subTreeSize <= 4 {
+		return errors.New(`the subTree must at least contain 4 bytes for the height
+		plus some data for the sum`)
+	}
+	for {
+		subTree := make([]byte, subTreeSize)
+		_, readErr := io.ReadFull(r, subTree)
+		if readErr == io.EOF {
+			// All data has been read.
+			break
+		} else if readErr != nil {
+			return readErr
+		}
+		// The first 4 bytes of the subTree are mapped to a height in range
+		// [0,50].
```

Comment:
50 is probably too high, makes it much less likely that you get trees of repeating, incrementing, decrementing height. The fuzzer will actually do a pretty good job of making this happen, but I'm still thinking we should drop the max height we fuzz to 10 or so.

How much code could potentially be covered by allowing a height of 11 (or higher) as opposed to just stopping at 10? The code is very simple, I'm guessing that even if the max height was 4 we'd still be able to cover all cases.

Because the fuzzer does a probabilistic search, we want to make sure we're giving it the tiniest search spaces that still cover all code cases, so that the fuzzer can spend it's time searching for the interesting use cases instead of getting lost in uninteresting cases.